fed/cleanup.py

```diff
@@ -23,97 +23,86 @@
+        the `graceful_stop()` will be invoked as well and the checking thread will be
+        notifiled to exit gracefully.
+
+        B. If the main thread are broken before sending the notification flag to the
+        sending thread, the monitor thread will detech that and it joins until the main
+        thread exited, then notifys the checking thread.
+    """
+
+    def __init__(self) -> None:
+        # `deque()` is thread safe on `popleft` and `append` operations.
+        # See https://docs.python.org/3/library/collections.html#deque-objects
+        self._sending_obj_refs_q = deque()
+        self._check_send_thread = None
+        self._monitor_thread = None
+        self._lock_on_sending_q = threading.Lock()
+        self._lock_on_send_thread = threading.Lock()
+
+    def start(self, exit_when_failure_sending=False):
+        self._exit_when_failure_sending = exit_when_failure_sending
+
+        def __check_func():
+            self._check_sending_objs()
+
+        self._check_send_thread = threading.Thread(target=__check_func)
+        self._check_send_thread.start()
         logger.info('Start check sending thread.')
 
-        global _monitor_thread
-        if not _monitor_thread:
-            _monitor_thread = threading.Thread(target=_main_thread_monitor)
-            _monitor_thread.start()
-            logger.info('Start check sending monitor thread.')
-
-
-def push_to_sending(obj_ref: ray.ObjectRef):
-    _start_check_sending()
-    global _sending_obj_refs_q
-    _sending_obj_refs_q.append(obj_ref)
-
-
-def notify_to_exit():
-    # Sending the termination signal
-    push_to_sending(True)
-    logger.info('Notify check sending thread to exit.')
-
-
-def wait_sending():
-    global _check_send_thread
-    if _check_send_thread:
-        notify_to_exit()
-        _check_send_thread.join()
+        def _main_thread_monitor():
+            main_thread = threading.main_thread()
+            main_thread.join()
+            self._notify_to_exit()
+
+        self._monitor_thread = threading.Thread(target=_main_thread_monitor)
+        self._monitor_thread.start()
+        logger.info('Start check sending monitor thread.')
+
+    def graceful_stop(self):
+        assert self._check_send_thread is not None
+        self._notify_to_exit()
+        self._check_send_thread.join()
+
+    def _notify_to_exit(self):
+        # Sending the termination signal
+        self.push_to_sending(True)
+        logger.info('Notify check sending thread to exit.')
+
+    def push_to_sending(self, obj_ref: ray.ObjectRef):
+        self._sending_obj_refs_q.append(obj_ref)
+
+    def _check_sending_objs(self):
+        def _signal_exit():
+            os.kill(os.getpid(), signal.SIGTERM)
+
+        assert self._sending_obj_refs_q is not None
+
+        while True:
+            try:
+                obj_ref = self._sending_obj_refs_q.popleft()
+            except IndexError:
+                time.sleep(0.5)
```

Comment:
Maybe we can reduce the sleep interval, say 0.1 seconds?

Reply:
DONE